commands/command_track.go

```diff
@@ -41,50 +42,110 @@ func trackCommand(cmd *cobra.Command, args []string) {
 	}
 
 	lfs.InstallHooks(false)
-	knownPatterns := findPatterns()
+	knownPatterns := git.GetAttributePaths(config.LocalWorkingDir, config.LocalGitDir)
 
 	if len(args) == 0 {
 		Print("Listing tracked patterns")
 		for _, t := range knownPatterns {
-			Print("    %s (%s)", t.Pattern, t.Source)
+			if t.Lockable {
+				Print("    %s [lockable] (%s)", t.Path, t.Source)
+			} else {
+				Print("    %s (%s)", t.Path, t.Source)
+			}
 		}
 		return
 	}
 
-	addTrailingLinebreak := needsTrailingLinebreak(".gitattributes")
-	attributesFile, err := os.OpenFile(".gitattributes", os.O_RDWR|os.O_APPEND|os.O_CREATE, 0660)
-	if err != nil {
-		Print("Error opening .gitattributes file")
-		return
-	}
-	defer attributesFile.Close()
-
-	if addTrailingLinebreak {
-		if _, werr := attributesFile.WriteString("\n"); werr != nil {
-			Print("Error writing to .gitattributes")
-		}
-	}
-
 	wd, _ := os.Getwd()
 	relpath, err := filepath.Rel(config.LocalWorkingDir, wd)
 	if err != nil {
 		Exit("Current directory %q outside of git working directory %q.", wd, config.LocalWorkingDir)
 	}
 
+	changedAttribLines := make(map[string]string)
+	var readOnlyPatterns []string
+	var writeablePatterns []string
 ArgsLoop:
 	for _, unsanitizedPattern := range args {
 		pattern := cleanRootPath(unsanitizedPattern)
 		for _, known := range knownPatterns {
-			if known.Pattern == filepath.Join(relpath, pattern) {
+			if known.Path == filepath.Join(relpath, pattern) &&
+				((trackLockableFlag && known.Lockable) || // enabling lockable & already lockable (no change)
+					(trackNotLockableFlag && !known.Lockable) || // disabling lockable & not lockable (no change)
+					(!trackLockableFlag && !trackNotLockableFlag)) { // leave lockable as-is in all cases
 				Print("%s already supported", pattern)
 				continue ArgsLoop
 			}
 		}
 
-		// Make sure any existing git tracked files have their timestamp updated
-		// so they will now show as modifed
-		// note this is relative to current dir which is how we write .gitattributes
-		// deliberately not done in parallel as a chan because we'll be marking modified
+		// Generate the new / changed attrib line for merging
+		encodedArg := strings.Replace(pattern, " ", "[[:space:]]", -1)
+		lockableArg := ""
+		if trackLockableFlag { // no need to test trackNotLockableFlag, if we got here we're disabling
+			lockableArg = " " + git.LockableAttrib
+		}
+
+		changedAttribLines[pattern] = fmt.Sprintf("%s filter=lfs diff=lfs merge=lfs -text%v\n", encodedArg, lockableArg)
+
+		if trackLockableFlag {
+			readOnlyPatterns = append(readOnlyPatterns, pattern)
+		} else {
+			writeablePatterns = append(writeablePatterns, pattern)
+		}
+
+		Print("Tracking %s", pattern)
+
+	}
+
+	// Now read the whole local attributes file and iterate over the contents,
+	// replacing any lines where the values have changed, and appending new lines
+	// change this:
+
+	attribContents, err := ioutil.ReadFile(".gitattributes")
```

Comment:
Yes, because I'm overwriting the same file. I can't progressively read and overwrite it at the same time 😄 I could write a new attributes file then delete & rename, but it's more complex and I prefer simpler. I recognise that this way is heavier on memory usage, but `.gitattributes` files are never going to be massive (even a 1MB attributes file would be unthinkable) so it's a not really an issue.